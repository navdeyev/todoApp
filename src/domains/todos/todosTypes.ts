export enum TodoStatus {
    PENDING = 'PENDING',
    IN_PROGRESS = 'IN_PROGRESS',
    COMPLETE = 'COMPLETE',
}

export interface ITodoStep {
    id: string,
    title: string,
    details: string
}

export interface ITodo {
    id: string,
    title: string,
    goal: string,
    steps: ITodoStep[],
    status: TodoStatus
}
