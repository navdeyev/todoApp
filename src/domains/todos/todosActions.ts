import {ActionCreator, AnyAction} from 'redux';

import {ITodo} from 'domains/todos/todosTypes';

export enum TodosActions {
    LOAD_TODOS = 'LOAD_TODOS',
    LOAD_TODOS_PENDING = 'LOAD_TODOS_PENDING',
    LOAD_TODOS_SUCCESS = 'LOAD_TODOS_SUCCESS',
    LOAD_TODOS_ERROR = 'LOAD_TODOS_ERROR',

    SELECT_TODO = 'SELECT_TODO',

    UPDATE_STATUS = 'UPDATE_STATUS',
    UPDATE_STATUS_PENDING = 'UPDATE_STATUS_PENDING',
    UPDATE_STATUS_SUCCESS = 'UPDATE_STATUS_SUCCESS',
    UPDATE_STATUS_ERROR = 'UPDATE_STATUS_ERROR',
}

export interface ILoadTodosSuccessAction extends AnyAction {
    payload: ITodo[];
}

const loadTodos: ActionCreator<AnyAction> = () => {
    return {type: TodosActions.LOAD_TODOS};
};

const loadTodosPending: ActionCreator<AnyAction> = () => {
    return {type: TodosActions.LOAD_TODOS_PENDING};
};

const loadTodosSuccess: ActionCreator<ILoadTodosSuccessAction> = (todos: ITodo[]) => {
    return {
        payload: todos,
        type: TodosActions.LOAD_TODOS_SUCCESS
    };
};

const loadTodosError: ActionCreator<AnyAction> = () => {
    return {type: TodosActions.LOAD_TODOS_ERROR};
};

const selectTodo: ActionCreator<AnyAction> = (todoId: string) => {
    return {
        payload: todoId,
        type: TodosActions.SELECT_TODO
    };
};

const updateTodoStatus: ActionCreator<AnyAction> = (todoId: string) => {
    return {
        payload: todoId,
        type: TodosActions.UPDATE_STATUS
    };
};

const updateTodoStatusPending: ActionCreator<AnyAction> = () => {
    return {type: TodosActions.UPDATE_STATUS_PENDING};
};

const updateTodoStatusSuccess: ActionCreator<ILoadTodosSuccessAction> = (todos: ITodo[]) => {
    return {
        payload: todos,
        type: TodosActions.UPDATE_STATUS_SUCCESS
    };
};

const updateTodoStatusError: ActionCreator<AnyAction> = () => {
    return {type: TodosActions.UPDATE_STATUS_ERROR};
};

export default {
    loadTodos,
    loadTodosError,
    loadTodosPending,
    loadTodosSuccess,
    selectTodo,
    updateTodoStatus,
    updateTodoStatusError,
    updateTodoStatusPending,
    updateTodoStatusSuccess
}