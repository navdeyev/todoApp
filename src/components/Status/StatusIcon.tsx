import styled from 'components/Styled/styledComponents';
import {defaultTheme} from 'components/Styled/themes';
import {rem} from 'components/Styled/utils';
import {TodoStatus} from 'domains/todos/todosTypes';

import complete from './complete.png';
import inProgress from './in_progress.png';
import pending from './pending.png';

import {ITodoStatusProps} from './Status';

export const getStatusIcon = (status: TodoStatus) => {
    if (TodoStatus.COMPLETE === status) {
        return complete;
    }

    if (TodoStatus.IN_PROGRESS === status) {
        return inProgress;
    }
    return pending;
};

const StatusIcon = styled.div`
  cursor: pointer;
  width: ${ rem(28) };
  height: ${ rem(28) };
  background-size: cover;
  background-repeat: no-repeat;
  background-position: center center;
  background-image: url(${ (props: ITodoStatusProps) => getStatusIcon(props.status) });
`;

StatusIcon.defaultProps = {theme: defaultTheme};
StatusIcon.displayName = 'StatusIcon';

export default StatusIcon;