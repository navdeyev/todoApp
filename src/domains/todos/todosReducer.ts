import * as R from 'ramda';
import {combineReducers} from 'redux';
import {getType} from 'typesafe-actions';

import {LoadingStates} from 'domains/loadingStates';
import {AppAction} from 'domains/rootReducer';
import {ITodo} from 'domains/todos/todosTypes';

import todoActions from './todosActions';

export const todos = (state: ITodo[] = [], action: AppAction) => {
    if (action.type === getType(todoActions.loadTodosSuccess) ||
        action.type === getType(todoActions.updateTodoStatusSuccess)) {
        return action.payload;
    }
    return state;
};

export const loadingState = (state = LoadingStates.NOT_STARTED, action: AppAction) => {
    const pendingArray = [
        getType(todoActions.loadTodosPending),
        getType(todoActions.updateTodoStatusPending),
    ];
    if (R.contains(action.type, pendingArray)) {
        return LoadingStates.LOADING;
    }

    const completeArray = [
        getType(todoActions.loadTodosSuccess),
        getType(todoActions.loadTodosError),
        getType(todoActions.updateTodoStatusSuccess),
        getType(todoActions.updateTodoStatusError)
    ];
    if (R.contains(action.type, completeArray)) {
        return LoadingStates.COMPLETE;
    }

    return state;
};

export const selectedTodoId = (state = '', action: AppAction) => {
    if (action.type === getType(todoActions.selectTodo)) {
        return state === action.payload ? '' : action.payload;
    }
    return state;
};

export default combineReducers({
    loadingState,
    selectedTodoId,
    todos,
});

