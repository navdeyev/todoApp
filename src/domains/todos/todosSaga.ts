import {call, put, takeLatest} from 'redux-saga/effects';

import {IApiService} from "service/api";
import {IServiceMap} from 'service/services';
import {loadTodosError, loadTodosPending, loadTodosSuccess, TodosActions} from './todosActions';

export function* loadTodos(apiService: IApiService) {
    yield put(loadTodosPending());

    try {
        const todos = yield call(apiService.loadTodoList);
        yield put(loadTodosSuccess(todos));
    } catch (e) {
        yield put(loadTodosError());
    }
}

function* todosSaga(serviceMap: IServiceMap) {
    yield takeLatest(TodosActions.LOAD_TODOS, loadTodos, serviceMap.apiService);
}

export default todosSaga;
