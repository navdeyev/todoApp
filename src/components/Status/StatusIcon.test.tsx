import {shallow} from 'enzyme';
import * as React from 'react';

import {TodoStatus} from 'domains/todos/todosTypes';
import StatusIcon from './StatusIcon';

describe('StatusIcon', () => {
    it('renders a component', () => {
        const props = {
            status: TodoStatus.COMPLETE,
            onClick: jest.fn()
        };
        expect(shallow(<StatusIcon {...props} />).getElement()).toMatchSnapshot();
    });
});
