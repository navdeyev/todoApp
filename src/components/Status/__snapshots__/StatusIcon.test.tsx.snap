// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusIcon renders a component 1`] = `
<div
  className="sc-bdVaJa jtBSla"
  onClick={[MockFunction]}
/>
`;
